RFC.md

```diff
@@ -0,0 +1,413 @@
+        attn_norm: Optional[nn.Module] = None,
+        mlp_norm: Optional[nn.Module] = None,
+        # New scale modules to allow scaling/gating attn output
+        attn_scale: Optional[nn.Module] = None,
+        mlp_scale: Optional[nn.Module] = None,
+    ) -> None:
+        super().__init__()
+        self.attn = attn
+        self.mlp = mlp
+        self.attn_norm = attn_norm or nn.Identity()
+        self.mlp_norm = mlp_norm or nn.Identity()
+        self.attn_scale = attn_scale or nn.Identity()
+        self.mlp_scale = mlp_scale or nn.Identity()
+
+    def forward(
+        self,
+        x: Tensor,
+        *,
+        mask: Optional[Tensor] = None,
+        input_pos: Optional[Tensor] = None,
+        **kwargs, # encoder_input and encoder_mask is not used for this module
+    ) -> Tensor:
+	    # Identical to TransformerDecoderLayer with addition of *_scale
+        attn_out = self.attn(self.attn_norm(x), mask=mask, input_pos=input_pos)
+
+        h = self.attn_scale(attn_out) + x
+
+        mlp_out = self.mlp(self.mlp_norm(h))
+
+        out = h + self.mlp_scale(mlp_out)
+        return out
+```
+
+### TransformerCrossAttentionLayer
+New layer. Discussion in the comments.
+```python
+class TransformerCrossAttentionLayer(nn.Module):
+    def __init__(
+		...
+    ) -> None:
+        super().__init__()
+        assert (
+            attn.pos_embeddings is None
+        ), "Positions are not computed for encoder inputs"
+        ...
+
+    def _skip_mask(self, mask: Optional[Tensor]) -> Optional[Tensor]:
+	    """Some tokens in x may not attend to any encoder inputs
+        due to the cross attention mask (encoder_mask). This results in
+        a full row of the attention matrix being masked out.
+
+        In the example below, the word "the" is masked from every embedding.
+
+        .. code-block:: text
+
+            |emb||emb||emb|
+        |The| x    x    x
+        |red|      x
+        |car| x
+
+        This results in no inputs into the softmax layer which causes a NaN.
+        The skip mask removes the output of the attention module and
+        mlp resulting in the token being skipped.
+
+        """
+		# There is more discussion around this here
+		# https://gist.github.com/drisspg/547648ded500d078961b7a3b3f11c310
+		# We might be able to optimize this approach more
+        if mask is None:
+            return None
+        if mask.dtype == torch.bool:
+            mask = ~mask
+        else:
+            mask = torch.isneginf(mask)
+        mask = torch.all(mask, dim=-1, keepdim=True)
+        return mask
+
+    def forward(
+        self,
+        x: Tensor,
+        *,
+        encoder_input: Optional[Tensor] = None,
+        encoder_mask: Optional[Tensor] = None,
+        input_pos: Optional[Tensor] = None,
+        **kwargs,
+    ) -> Tensor:
+        # Cross attention only layer (see red only) has no
+        # purpose if no conditional input is provided
+        if encoder_input is None:
+            return x
```

Comment:
I know why you decided to do this, just wanna point out that it's a bit unintuitive for generic cross-attention layer imo. We should make it abundantly clear that this is the behavior, otherwise people will get confused